client/v3/maintenance.go

```diff
@@ -213,40 +233,92 @@ func (m *maintenance) HashKV(ctx context.Context, endpoint string, rev int64) (*
 	return (*HashKVResponse)(resp), nil
 }
 
-func (m *maintenance) Snapshot(ctx context.Context) (io.ReadCloser, error) {
+func (m *maintenance) SnapshotWithVersion(ctx context.Context) (*SnapshotResponse, error) {
 	ss, err := m.remote.Snapshot(ctx, &pb.SnapshotRequest{}, append(m.callOpts, withMax(defaultStreamMaxRetries))...)
 	if err != nil {
 		return nil, toErr(ctx, err)
 	}
 
 	m.lg.Info("opened snapshot stream; downloading")
 	pr, pw := io.Pipe()
+
+	resp, err := ss.Recv()
+	if err != nil {
+		m.logAndCloseWithError(err, pw)
+	}
 	go func() {
+		// Saving response is blocking
+		err = m.save(resp, pw)
+		if err != nil {
+			m.logAndCloseWithError(err, pw)
+			return
+		}
 		for {
 			resp, err := ss.Recv()
 			if err != nil {
-				switch err {
-				case io.EOF:
-					m.lg.Info("completed snapshot read; closing")
-				default:
-					m.lg.Warn("failed to receive from snapshot stream; closing", zap.Error(err))
-				}
-				pw.CloseWithError(err)
+				m.logAndCloseWithError(err, pw)
 				return
 			}
+			err = m.save(resp, pw)
+			if err != nil {
+				m.logAndCloseWithError(err, pw)
+				return
+			}
+		}
+	}()
+	return &SnapshotResponse{
+		Header:   resp.Header,
+		Snapshot: &snapshotReadCloser{ctx: ctx, ReadCloser: pr},
+		Version:  resp.Version,
+	}, err
+}
 
-			// can "resp == nil && err == nil"
-			// before we receive snapshot SHA digest?
-			// No, server sends EOF with an empty response
-			// after it sends SHA digest at the end
+func (m *maintenance) Snapshot(ctx context.Context) (io.ReadCloser, error) {
+	ss, err := m.remote.Snapshot(ctx, &pb.SnapshotRequest{}, append(m.callOpts, withMax(defaultStreamMaxRetries))...)
```

Comment:
Can it call SnapshotWithVersion ?

Reply:
Not without changing behavior for canceled context. `TestMaintenanceSnapshotError` validates what happens if context is closed before reading from`ReaderCloser`. This bahavior changes in `SnapshotWithVersion`.

`TestMaintenanceSnapshotError` cancels context after connection is established, but before we read first chunk in stream. This means that ctx cancel is propagated to ReaderCloser before we read anything from `ReaderCloser`.

On the other hand `SnapshotWithVersion` waits for first response from stream is received, as it needs it to get information about version. This means that first chunk of snapshot is already available for reading without returning error due to ctx cancelled. 

Aside of needing to drop this test, I think that difference in blocking between `Snapshot` and `SnapshotWithVersion` has enough impact that we should keep them separate.